[Tracker Issue] Add Prometheus Remote Write compatibility tests to OTel Collector

Use the test suite at https://github.com/tomwilkie/remote-write-compliance by @tomwilkie for compatibility testing of the Prometheus Remote Write Exporter in the Collector.

Note: Compatibility tests status as of June 9 2021:

Update on Compliance tests, Collector 0.28.0:
    --- FAIL: TestRemoteWrite/otelcollector (0.01s)
        --- FAIL: TestRemoteWrite/otelcollector/RepeatedLabels (10.01s) -- (Emmanuel working on this, filed https://github.com/open-telemetry/opentelemetry-collector/issues/3407 as it is an issue with receiver/prometheus) PR available at https://github.com/open-telemetry/opentelemetry-collector/pull/3408
        --- PASS: TestRemoteWrite/otelcollector/JobLabel (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/Counter (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/Summary (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/InstanceLabel (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/HonorLabels (10.01s)
        --- FAIL: TestRemoteWrite/otelcollector/Retries500 (10.01s)
                        FYI: https://github.com/prometheus/compliance/pull/31
                        OTel is using https://github.com/cenkalti/backoff
        --- PASS: TestRemoteWrite/otelcollector/EmptyLabels (10.01s)
        --- FAIL: TestRemoteWrite/otelcollector/Up (10.01s)
                       Will be addressed by https://github.com/open-telemetry/opentelemetry-collector/pull/3116 
        --- PASS: TestRemoteWrite/otelcollector/Histogram (10.01s)
        --- FAIL: TestRemoteWrite/otelcollector/Invalid (10.01s)
                       Will be addressed by https://github.com/open-telemetry/opentelemetry-collector/pull/3116 
        --- PASS: TestRemoteWrite/otelcollector/NameLabel (10.01s)
        --- FAIL: TestRemoteWrite/otelcollector/Staleness (10.01s)
                       This needs to be addressed
        --- PASS: TestRemoteWrite/otelcollector/Headers (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/SortedLabels (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/Retries400 (10.01s)
        --- PASS: TestRemoteWrite/otelcollector/Timestamp (10.02s)
        --- PASS: TestRemoteWrite/otelcollector/Gauge (10.02s)
        --- PASS: TestRemoteWrite/otelcollector/Ordering (11.15s)

--- a/.github/workflows/prometheus-compliance-tests.yml
+++ b/.github/workflows/prometheus-compliance-tests.yml
@@ -1,0 +1,51 @@
+name: prometheus-compliance-tests
+on:
+  push:
+    branches: [ main ]
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+*'
+  pull_request:
+
+jobs:
+  prometheus-compliance-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+      - name: Cache Tools
+        id: tool-cache
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-tool-binaries
+        with:
+          path: /home/runner/go/bin
+          key: v1-tools-${{ runner.os }}-${{ hashFiles('./internal/tools/go.mod') }}
+      - name: Setup Go
+        uses: actions/setup-go@v2.1.4
+        with:
+          go-version: 1.17
+      - name: Setup Go Environment
+        run: |
+          echo "GOPATH=$(go env GOPATH)" >> $GITHUB_ENV
+          echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
+      - name: Cache Go
+        id: module-cache
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-go-modules
+        with:
+          path: |
+            /home/runner/go/pkg/mod
+            /home/runner/.cache/go-build
+          key: v1-go-pkg-mod-${{ runner.os }}-${{ hashFiles('**/go.mod', '**/go.sum') }}
+      - run: make otelcontribcol
+      - name: Checkout compliance repo
+        uses: actions/checkout@v2
+        with:
+          repository: prometheus/compliance
+          path: compliance
+      - name: Copy binary to compliance directory
+        run: mkdir compliance/remote_write/bin && cp ./bin/otelcontribcol_linux_amd64 compliance/remote_write/bin/otelcol_linux_amd64
+      - name: Run compliance tests
+        run: go test --tags=compliance -run "TestRemoteWrite/otel/.+" -v ./
+        working-directory: compliance/remote_write

